### ISSUE TITLE
[RFC] .env environment support

### ISSUE THREAD

**Issue by timneutkens** [zeit/next.js#11106]
## Goals

- Provide a secure way to access environment variables in a granular way for pages
- Add dotenv (`.env`) support with the same `.env` file loading semantics as Create React App: https://create-react-app.dev/docs/adding-custom-environment-variables

## Background

Over the last few months we've had a lot of feedback from companies running into confusion around how to handle environment variables.

The main feedback has been mostly:

- It's unclear how / where to load environment variables in development.
    - For this reason most tend to fall back to using `dotenv` (almost 50% of Next.js apps).
- It's unclear when / if a environment variable is exposed to the browser
- You can't limit the exposed environment variables to pages
- It's unclear how to expose environment variables to the browser

## Proposal

The main takeaway from the feedback has been that `process.env` is too global and it's not clear when it's exposed to the browser or inlined into the JavaScript bundles.

In order to provide a clear path from Server → Browser we can leverage the new data fetching methods introduced in Next.js 9.3: `getStaticProps`, `getStaticPaths`, and `getServerSideProps`. These methods always run in the Node.js environment and never in the browser. Which should clear up quite a lot of confusion.

We can do this by exposing an `env` parameter to the methods:

```js
// pages/index.js
import fetchData from '../lib/api'

export async function getStaticProps({ env }) {
  const data = await fetchAPI('https://example.com/api/v1/posts', {
    token: env.ACCESS_TOKEN
  })
  return {
    props: {
      data
    }
  }
}

export default function HomePage({ data }) {
  return <div>
    <h1>{data.title}</h1>
  </div>
}
```

This would also cover the Server → Browser handoff, as an environment variable could be returned in the data fetching methods to expose it:

```js
// pages/index.js
import fetchData from '../lib/api'

export async function getStaticProps({ env }) {
  return {
    props: {
      token: env.ACCESS_TOKEN
    }
  }
}

export default function HomePage({ token }) {
  return <div>
    <h1>This value was exposed to the browser {token}</h1>
  </div>
}
```

Instead of all environment variables being exposed though you provide which variables have to be injected. This ensures pages only get access to the environment variables needed for that page:

```js
// pages/index.js
import fetchData from '../lib/api'

export const config = {
  env: ['ACCESS_TOKEN']
}

export async function getStaticProps({ env }) {
  const data = await fetchAPI('https://example.com/api/v1/posts', {
    token: env.ACCESS_TOKEN
  })
  return {
    props: {
      data
    }
  }
}

export default function HomePage({ data }) {
  return <div>
    <h1>{data.title}</h1>
  </div>
}
```

For API routes the environment variables will be exposed on `req.env`:

```js
// pages/api/api-route.js
export const config = {
  env: ['ACCESS_TOKEN']
}


export default function ApiRoute(req, res) {
  console.log(req.env.ACCESS_TOKEN) // Logs out the ACCESS_TOKEN variable 
  return res.json({ hello: 'world' })
}
```

Environment variables will be loaded from `.env` files, using the same loading mechanism as Create React App: https://create-react-app.dev/docs/adding-custom-environment-variables

For example:

```
# .env.local
ACCES_TOKEN=abcdefgh
```

Will set `ACCESS_TOKEN` to `abcdefgh` be used in pages.

---

Overall these changes should simplify usage of environment variables for the majority of use cases while improving security.

**Comment by laugharn** [comment]
By default Now excludes .env files from deployments, would there be a plan to change that when/if this RFC lands?

**Comment by timneutkens** [comment]
@laugharn no, when deploying you would pre-define the environment variables in the ZEIT dashboard for the project (this is currently being added). Uploading a `.env` file would be less secure as these can't be encrypted whereas secrets are.

**Comment by eric-burel** [comment]
That sounds quite targeted at pages. 
What about config in other parts of the client side code. Say for example that I want to set my Apollo client in a certain way or configure a domain-level lib with some settings?

That's kind of similar to an issue I have with React or Redux more broadly, when a `Provider` component is containing too much logic: the object it produces are only available within React components, while advanced use cases might require an usage outside React.

I lack clearer example but you get the point, you may want to clarify patterns about how to access those variables outside pages.

**Comment by timneutkens** [comment]
Those patterns are already covered by the `env` key still. We might extend it to take in just an array of variables that have to be global. However keep in mind that these will be inlined at build time. Which is kind of expected as they're global and exposed client-side.

**Comment by eric-burel** [comment]
You meant that this an additional feature, but the `env` key behaviour will still coexist?
But then I am not sure it address the issue I've opened about public config https://github.com/zeit/next.js/issues/11131.
In the end you may still leak variables from the `env` key if you still choose this pattern or if you are forced too because you want settings outside of pages.

**Comment by timneutkens** [comment]
> In the end you may still leak variables from the env key if you still choose this pattern or if you are forced too because you want settings outside of pages.

We'll just add extra docs on the way that it should be used. Even currently it's for *exposing env variables everywhere* including client-side.

**Comment by eric-burel** [comment]
Sounds good to me if patterns are well documented. Isolating client side variables in `process.env.public` sounds sufficient to me for example.

> reading it in API routes through normal process.env
Normal `process.env` is very limited though, no nesting, no arrays, no defaults. I'd rather have everything clearly defined in `next.config.js`, even when the `env` key then calls `process.env`. Same applies to `dotenv` btw (I can be wrong though I did not use it a lot).

Not directly related but I've noticed that the `env` variables defined in the config are not available when starting a custom server. 
That probably makes total sense since it's a custom server, but do you have an idea how we could "force access" to the `next.config` env variables at this point? Or should I fallback to calling `process.env`?

To sum it up, I would expect a chain like this:
`.dotenv > process.env > next.config.js > custom server, api routes, and client code`. The `next.config.js` `env` key being the single source of truth for config, being env based or not, except only for runtime config.

**Comment by eric-burel** [comment]
Use cases for `runtime` config are still unclear to me btw. I feel like a confusion between the concept of "app settings" and "environment variables". In Meteor app we are used to abstract away everything and just have settings, but they are runtime only.

**Comment by borekb** [comment]
Just going to post a piece from our internal docs – not yet sure how relevant it is to this RFC and also, we have room for improvement, but anyway:

---

<img width="995" alt="Screenshot 2020-03-17 at 10 56 38" src="https://user-images.githubusercontent.com/101152/76844721-00dd8500-683e-11ea-94ea-f91fe258aac8.png">

---

<img width="995" alt="Screenshot 2020-03-17 at 10 56 17" src="https://user-images.githubusercontent.com/101152/76844728-02a74880-683e-11ea-8fc7-17bef226e326.png">

**Comment by eric-burel** [comment]
Last comment for me: TS typings are wrong for `process.env` when using nested objects. Env variables are supposed to be string, but `next.config.js` `env` key will allow nested object and arrays (which is nice).

```
const whitelist =  process.env.apolloServer.cors.whitelist // error because apolloServer is supposed to be a string
```
@borekb Totally related to https://github.com/zeit/next.js/issues/11131

I think in the meantime I'll fallback to runtime config, as my app is not really static.

**Comment by eric-burel** [comment]
To sum it up: I would expect the `build` time config of `env` to behave exactly like the runtime config, syntax-wise. I understand the technical difference of both but not the syntactical difference. 

Using `dotenv` could be included as a default in Next but only affects the content of `process.env`. 

Instead of using CRA prefixing for env variable that should also be available client-side, (honestly I hate it, it's not intuitive) we could just expect the user to call `process.env` when defining its client side settings, and then access this settings through Next methods or `getSetting` method or whatever.

Basically a Meteor-like pattern but that works at build time and with an even less confusing syntax (clearer separation of client-side and server-side settings).

**Comment by timneutkens** [comment]
> Last comment for me: TS typings are wrong for `process.env` when using nested objects. Env variables are supposed to be string, but `next.config.js` `env` key will allow nested object and arrays (which is nice).
> 
> ```
> const whitelist =  process.env.apolloServer.cors.whitelist 
> // error because apolloServer is supposed to be a string
> ```
> 
> @borekb Totally related to #11131
> 
> I think in the meantime I'll fallback to runtime config, as my app is not really static.

The types are not wrong though, you shouldn't nest objects in `env`, it's a coincidence that it works and it's not tested in our test suite.

**Comment by jesstelford** [comment]
I use [`dotenv-safe`](https://github.com/rolodato/dotenv-safe) heavily over `dotenv` as it provides useful errors when expected environment variables are missing (particularly useful when deploying to a new environment when you might have forgotten to add an env var).

It also enforces The use of a `.env.example` which means onboarding a new developer (or nuking your code directory and setting up again) is as easy as `cp .env.example .env` then replace the examples with real values, no more reading seperate docs or guessing what env vars the codebase requires.

Do you plan to support such a feature, or at least allow opting in to it?,

**Comment by iaincollins** [comment]
I'd chime in with some challenges around this that might be worth thinking about for you to ignore if you feel they are not helpful. :-)

1. It would be helpful on a projects to be able to know what env vars *can* / need to be configured.

I run into this pain point a fair bit, including on old projects I haven't looked at in a while and have refactored without updating all the documentation.

It would be great to design a mechanism that could display this using reflection (for example at start up, maybe with displaying default values where possible, like `APP_URL=http://localhost:3000`?) or perhaps if people were required to list them somewhere (as with `now.json` on now.sh) that could be really helpful when on-boarding folks on new projects.

2. It would be helpful to be able to clearly define which env vars are protected and which should be exported client side.

I feel like this is a fairly common pitfall. Having a mechanism whereby it's easy to categorically say "this env var is not safe to export to the client" (or better yet, explicitly having to say "this *is* safe to export to the client") and have that enforced when bundling, even if someone then tries to access it client side flow, would be valuable.

I get that the new data fetching methods help avoid this, but also appreciate there is a lot to take in for new developers and the nature of Next.js (making it easier to do more complex things) means people can run into complexity sooner and get overwhelmed by everything they are trying to take on board at once and so miss things like this.

If not feasible, perhaps using reflection to at least warn people what env vars are being exported to the client at dev run time / build time is possible?

**Comment by dwest-teo** [comment]
This seems like it could be very helpful for cleaning up build processes as well as providing more granular control of what gets sent down to the client.  Presently, we have to build a Docker image for each environment we deploy to in order to ensure env vars are included at build time.  It seems to me that the feature described here would enable one to build a single Docker image for all environments and then consume the appropriate env vars at runtime.

With that said, can we provide a way to have more granular control over which `.env` file is loaded?  Could we specify an environment when starting the app as a CLI parameter perhaps?  If so, how could this work with a custom server?

**Comment by iaincollins** [comment]
```
With that said, can we provide a way to have more granular control over which .env file is loaded? Could we specify an environment when starting the app as a CLI parameter perhaps? If so, how could this work with a custom server?
```

While I get this us easy to put under a nice-to-have, it would be really nice and would second this.

In practice, I've found being able to easily load configs for different environments super handy.

**Comment by kachkaev** [comment]
I'm wondering how the RFC is planning to deal with TS typings for `env.SOMETHING` and `export const config = {env: ['SOMETHING']}`. These typings are important for a good DX, especially in large projects  🤔

Also curious how to deal with parsing and validation of env variables. E.g. `env.ENABLE_SOMETHING` can be equal to `"1"`, `"true"`, `""`, `undefined`, `"0"`, `"false"` and `"anything invalid"`. Parsing this value in each component with some wrapper function seems like extra work, which is prone to errors. Ideally, you want to get the validated values in one place and then feed the rest of your app with already validated and transformed values.

I've been using [`envalid`](https://www.npmjs.com/package/envalid) in quite a few projects and it solved the above issues for my teams. Wondering if we could use it in Next.js env in some way too. E.g.

```js
// env.config.js
const envalid = require("envalid")

const env = envalid.cleanEnv(
  process.env,
  {
    WHATEVER: envalid.str({ // app won't start if this is not defined
      desc: "Some required string value",
    }),
    ENABLE_SOMETHING: envalid.bool({ // app will start with ENABLE_SOMETHING=1, but wont start with ENABLE_SOMETHING=42
      default: true,
      devDefault: false,
      desc: "Some boolean that is true by default in prod, but false in dev",
    }),
    SOME_HOST: envalid.host({ // app won't start if SOME_HOST=not/a/valid/host
      default: "localhost",
      desc: "Host to connect to for fetching some data",
    }),
  },
  { strict: true }, // does not pass 
)

module.exports = env // result is typed with TS if checkJs is enabled in tsconfig.json
```

The _typeof env imported from env.config.js_ can be used in some .d.ts file, which will inform TS about property keys and typings within `env`. WDYT?

**Comment by blainegarrett** [comment]
Would this follow the 12 factor apps methodology with respect to CI/CD? i.e. the ability to create one build and deploy to different environments (qa, staging, prod)?  At my last job, this was a pain point in our build setup - on release, we would have to create 3 separate builds of our Next app. Granted, we were not hosting on now.sh

**Comment by borekb** [comment]
@blainegarrett We build a single Docker image for multiple requirements but have to avoid all the [built-in support](https://nextjs.org/docs/api-reference/next.config.js/environment-variables) for environment variables in Next.js, like noted [above](https://github.com/zeit/next.js/issues/11106#issuecomment-599981012).
